### ISSUE TITLE
Datasets for train_actor_rl.sh

### ISSUE THREAD

**Issue by doviettung96** [salesforce/CodeRL#24]
Dear @henryhungle ,
According to the script `train_actor_rl.sh` and `trainer_rl.py`, we have to prepare several JSON/pickle files like `gen_solutions_critic_scores.pkl` and/or `baseline_solutions.json`. Also, I found that we don't have `gen_solutions.json` as well. How can I generate them?
I found that we can generate `gen_solutions_critic_scores.pkl` by running `generate_critic_scores.sh` without the `gt_solutions` flag. However, that one requires `gen_solutions.json` which I can't find any scripts that do the job. I got stuck at this.
I think we can generate `baseline_solutions` or `gen_solutions` just like the normal generating program procedure, right? 
Thanks.

**Comment by henryhungle** [comment]
@doviettung96  Yes correct! We can generate `baseline_solutions` and `gen_solutions` with the normal generation procedure. After that, you can run the script to obtain the critic scores for `gen_solutions` before fine-tuning the actor model.

**Comment by doviettung96** [comment]
Thank you @henryhungle . 
I have 1 more question: "Both of `baseline_solutions` and `gen_solutions` are generated from the same model, right? They are just different from each other in the decoding method (the former uses sampling while the latter use greedy/ beam search). Also, do we need to re-generate them after some training steps/ epochs? Since if we generate them only once before the training, it might not reflect correctly the current policy model.
Thanks.

**Comment by doviettung96** [comment]
Oh. Anyway, when I looked at `trainer_rl.py` to see the training loop, I couldn't find any generating-scoring steps (actor-critic) but only find a trained (fixed) critic, and the generated programs are loaded from dataloader (fixed). Is there any reason for that? Like stability? Or by experiment, you found that perform better? 
Thanks.

**Comment by henryhungle** [comment]
@doviettung96 thanks for your comments. Yes both `baseline_solutions` and `gen_solutions` are from the same model. They were generated by different decoding methods. 
Yes we can actually generate them in multiple rounds (e.g. after each finetuning epoch). In our paper, we only generate in one round and let the model learn with synthetic samples and their estimate rewards.

**Comment by henryhungle** [comment]
> Oh. Anyway, when I looked at `trainer_rl.py` to see the training loop, I couldn't find any generating-scoring steps (actor-critic) but only find a trained (fixed) critic, and the generated programs are loaded from dataloader (fixed). Is there any reason for that? Like stability? Or by experiment, you found that perform better? Thanks.

actually the finetuning code can be further optimized so that in a pipeline, you can generate new samples, pass them to a critic and get the scores directly for each generated sample. We followed the prior RL best practice that often freezes the critic model while finetuning the actor.

**Comment by doviettung96** [comment]
So, in summary, you fixed your critic, and generate samples only before the training started. As a result, the estimated rewards for those generated samples are fixed as well. In the training loop, you only fine-tune the actor to optimize that estimated rewards. 
However, it you did not re-generate the samples and re-calculate the estimated rewards, how can you even optimize the actor? I mean that the RL-loss requires both of the return (which requires the rewards to be updated) and log-prob of the generated samples. In your case, both of them are fixed, so how can we make any changes to the actor?

**Comment by henryhungle** [comment]
Ideally we want to train the model in an online manner with updated rewards/samples at every training steps. However, in practice, as the rewards also depend on the execution results of the code samples, it will be tricky (e.g. executing some code samples may break the training script). In our case, given a round of synthetic sample generation, the model will still be optimized by penalizing the bad programs and rewarding more correct programs. Note that before RL finetuning, we warmup the model with conventional supervise finetuning so that the synthetic code quality is decent enough.

**Comment by doviettung96** [comment]
Yeah. The execution after generation is not just slow but also bug-full. The freezing critic and warm-up trick are quite practical. Thank you.

**Comment by doviettung96** [comment]
Hi @henryhungle,
Sorry for bothering again. This time, I can not find the right inputs for training critic. As you know, I generated `gen_solutions` with generate process and fill in the `code`, `result` of each generated program. However, I can not find anywhere with the `error_type`. That one is not a requirement for training Actor RL but it is for training Critic. 
Thanks.

**Comment by doviettung96** [comment]
Oh. FYI, I think I managed to fix that by finding the Exception class in the `error` from `test_one_solution`. If some error is new, I just add it to the list (mostly runtime errors).
